Introduce PlaybackControlRequest types

ericmlujan opened the pull request:
### Changelog
None

### Docs
None

### Description
Introduce new PlaybackControlRequest protocol message. Currently guarding behind the `unstable` ff and excluding from documentation.

jtbandes commented on rust/foxglove/src/websocket/ws_protocol/client/player_state.rs:
```diff
@@ -0,0 +1,144 @@
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum Playing {
```
yeah, I was pondering suggesting changing PlayerState to PlaybackState 🤔

ericmlujan commented on rust/foxglove/src/websocket/ws_protocol/client/player_state.rs:
```diff
@@ -0,0 +1,144 @@
+    pub seek_playback_time: Option<PlayerTime>,
+    /// Previous playback time (will be defined if a seek has occurred)
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub seek_previous_playback_time: Option<PlayerTime>,
```
Re this: I think we should remove this field and only provide the requested seek time, providing the other field feels unnecessary

ericmlujan commented on rust/foxglove/src/websocket/ws_protocol/client/player_state.rs:
```diff
@@ -0,0 +1,144 @@
+    pub seek_previous_playback_time: Option<PlayerTime>,
+}
+
+impl JsonMessage for PlayerState {}
```
Heard, I'm down with that. Thanks for linking that issue, it was helpful

ericmlujan commented:
@achim-k @jtbandes 
Thanks for the reviews, I went ahead and made some changes:
- Changed the underlying representation to being binary
- Introduced enum states for `Buffering` and `Ended` in the `PlaybackState` enum (no longer `Playing` 😅)
- Changed the seek time to be a primitive u64 and eliminated the redundant previous seek time field (we can add it in if needed). Still making it optional and signaling with a flag as to whether a seek occurred (in which case that field should be interpreted). Notably, I've erred towards serializing the seek time even if a seek hasn't occurred, to keep the message size constant. I figured 8 bytes was pretty minimal overhead to simplify parsing logic and allow for preallocating a fixed-size memory block.

gasmith commented on rust/foxglove/src/websocket/ws_protocol/client/player_state.rs:
```diff
@@ -0,0 +1,201 @@
+
+        let playback_speed = data.get_f32_le();
+        let had_seek = data.get_u8() != 0;
+        let seek_time = if had_seek && data.len() >= 8 {
```
Since L53 validates length:

```suggestion
        let seek_time = if had_seek {
```

achim-k commented on rust/foxglove/src/websocket/ws_protocol/client/player_state.rs:
```diff
@@ -0,0 +1,201 @@
+    /// Playback speed
+    pub playback_speed: f32,
+    /// Seek playback time in nanoseconds (only set if a seek has been performed)
+    pub seek_time: Option<u64>,
```
Should we also add the current time here? It kind of says "we are now playing from time X" or "at time X we are buffering" etc.

ericmlujan replied:
> ... add a field here indicating what is being played back?

I didn't think much about the name of the recording, I could see that being useful to send through ServerInfo. We'd then need UI support to display that name at the top of the viewport

gasmith replied:
I agree that we need a response from the server, to indicate that it has (or hasn't?) handled the request. In addition to the reason you describe, there's also the matter of this being a full-duplex channel: the client will be receiving messages concurrently while it waits for the server to handle the playback control request, and the client (presumably) needs to identify the cut-off between old messages (before the server handled the playback control request) and new messages.

Which means that we also need to guard against a faulty server implementation that fails to reply (in a timely manner). [We put some safeguards in place for service calls](https://github.com/foxglove/foxglove-sdk/blob/72a311a9abd475b3251796a5c6fc3afe319f8cef/rust/foxglove/src/websocket/service/response.rs#L81-L86), and we might want to considering doing the same here.

Which leads us to a question: Instead of making a new primitive ws_protocol message type, could this ride on top of the ServiceCall/ServiceResponse messages? Perhaps we introduce a new well-known service name (`/foxglove/playback_control`?) and schema?

gasmith commented on rust/foxglove/src/websocket/ws_protocol/client/playback_control_request.rs:
```diff
@@ -0,0 +1,201 @@
+    Playing = 0,
+    Paused = 1,
+    Buffering = 2,
+    Ended = 3,
```
I'm not sure how these would be used.

gasmith commented on rust/foxglove/src/websocket/ws_protocol/client/playback_control_request.rs:
```diff
@@ -0,0 +1,201 @@
+#[repr(u8)]
+pub enum PlaybackState {
+    Playing = 0,
+    Paused = 1,
```
Should this be phrased as a command? `Pause`, `Play`, etc.

gasmith commented on rust/foxglove/src/websocket/ws_protocol/client/playback_control_request.rs:
```diff
@@ -0,0 +1,201 @@
+    /// Playback state
+    pub playback_state: PlaybackState,
+    /// Playback speed
+    pub playback_speed: f32,
```
There's certainly an upper bound on playback speed. There may also be situations where the simulation is incapable at running at non-realtime speeds. I wonder whether that server needs to be able to communicate that to the client somehow, so that we can adjust the UI speed controls accordingly.

ericmlujan replied:
>Surfacing some notion of inferred playback speed in the UI would help the user to understand when the server can't keep up with the speed they're requesting. This might even make sense across the board (e.g., when playing back regular ol' recordings from the data platform).

I think this is a good idea regardless. Perhaps we could take the logic we use for displaying the topic Hz and repurpose it across all messages (I noticed when publishing a 1Hz topic in my local spike this also tracked the playback speed nicely with no explicit feedback).

ericmlujan commented on rust/foxglove/src/websocket/ws_protocol/client/playback_control_request.rs:
```diff
@@ -0,0 +1,201 @@
+    pub playback_speed: f32,
+    /// Seek playback time in nanoseconds (only set if a seek has been performed)
+    pub seek_time: Option<u64>,
+}
```
I think there will need to be a (separate, to be implemented) playback message from the server to the client that broadcasts that it's ended, or saying what state playback is in in regards to a PlaybackControlRequest, for example. I think the client should handle requesting a loop as a seek, once the server indicates that playback is over.

gasmith commented on rust/foxglove/src/websocket/ws_protocol/client/playback_control_request.rs:
```diff
@@ -0,0 +1,201 @@
+        let message = PlaybackControlRequest {
+            playback_command: PlaybackCommand::Pause,
+            playback_speed: 1.0,
+            seek_time: None,
```
Nit: Maybe provide Some seek time here, so we can snapshot that?

ericmlujan replied:
Resolved in [92baafa](https://github.com/foxglove/foxglove-sdk/pull/728/commits/92baafa5c8c81ef81309188cea2895bf368e5228)

gasmith commented on rust/foxglove/src/websocket/ws_protocol/client/playback_control_request.rs:
```diff
@@ -0,0 +1,201 @@
+    // + request_id_len (4 bytes)
+    // + request_id
+    fn parse_binary(mut data: &'a [u8]) -> Result<Self, ParseError> {
+        if data.len() < 1 + 4 + 1 + 8 {
```
Tack on 4 more for `request_id_len`

```suggestion
        if data.len() < 1 + 4 + 1 + 8 + 4 {
```

gasmith reviewed (approved):
LG, with some nits.
